src/Storages/MergeTree/MergeTreeIndexAnnoy.cpp

```diff
@@ -0,0 +1,323 @@
+    writeIntBinary(index->getNumOfDimensions(), ostr); // write dimension
+    index->serialize(ostr);
+}
+
+void MergeTreeIndexGranuleAnnoy::deserializeBinary(ReadBuffer & istr, MergeTreeIndexVersion /*version*/)
+{
+    uint64_t dimension;
+    readIntBinary(dimension, istr);
+    index = std::make_shared<AnnoyIndex>(dimension);
+    index->deserialize(istr);
+}
+
+
+MergeTreeIndexAggregatorAnnoy::MergeTreeIndexAggregatorAnnoy(
+    const String & index_name_,
+    const Block & index_sample_block_,
+    uint64_t number_of_trees_)
+    : index_name(index_name_)
+    , index_sample_block(index_sample_block_)
+    , number_of_trees(number_of_trees_)
+{}
+
+MergeTreeIndexGranulePtr MergeTreeIndexAggregatorAnnoy::getGranuleAndReset()
+{
+    // NOLINTNEXTLINE(*)
+    index->build(number_of_trees, /*number_of_threads=*/1);
+    auto granule = std::make_shared<MergeTreeIndexGranuleAnnoy>(index_name, index_sample_block, index);
+    index = nullptr;
+    return granule;
+}
+
+void MergeTreeIndexAggregatorAnnoy::update(const Block & block, size_t * pos, size_t limit)
+{
+    if (*pos >= block.rows())
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR,
+            "The provided position is not less than the number of block rows. Position: {}, Block rows: {}.",
+            toString(*pos), toString(block.rows()));
+
+    size_t rows_read = std::min(limit, block.rows() - *pos);
+    if (rows_read == 0)
+        return;
+
+    if (index_sample_block.columns() > 1)
+    {
+        throw Exception("Only one column is supported", ErrorCodes::LOGICAL_ERROR);
+    }
+
+    auto index_column_name = index_sample_block.getByPosition(0).name;
+    const auto & column_cut = block.getByName(index_column_name).column->cut(*pos, rows_read);
+    const auto & column_array = typeid_cast<const ColumnArray*>(column_cut.get());
+    if (column_array)
+    {
+        const auto & data = column_array->getData();
+        const auto & array = typeid_cast<const ColumnFloat32&>(data).getData();
+        if (array.empty())
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Array have 0 rows, but {} expected", rows_read);
+        const auto & offsets = column_array->getOffsets();
+        size_t num_rows = offsets.size();
+
+        /// All sizes are the same
+        size_t size = offsets[0];
+        for (size_t i = 0; i < num_rows - 1; ++i)
+        {
+            if (offsets[i + 1] - offsets[i] != size)
+            {
+                throw Exception(ErrorCodes::INCORRECT_DATA, "Arrays should have same length");
+            }
+        }
+        index = std::make_shared<AnnoyIndex>(size);
+
+        index->add_item(index->get_n_items(), array.data());
+        /// add all rows from 1 to num_rows - 1 (this is the same as the beginning of the last element)
+        for (size_t current_row = 0; current_row < num_rows; ++current_row)
+        {
+            index->add_item(index->get_n_items(), &array[offsets[current_row]]);
```

Comment:
```suggestion
        for (size_t current_row = 1; current_row < num_rows; ++current_row)
        {
            index->add_item(index->get_n_items(), &array[offsets[current_row - 1]]);
```

(it still does out-of-bounds read when adding the last element)